Add doc comments to tls config struct

dedemorton opened the pull request:
@ph OK, so here's a first pass at adding some doc comments. This simple exercise provoked some thoughts and questions. I'll add them here to capture my thoughts.

I didn’t add any example structs because I ran out of steam and figured you would want to add those.

I added the `yaml: ...` but not sure if what I did it right.

I'll work on the markdown (mdx) template tomorrow. I have to get some questions answered before proceeding.

Let me know if there is anything else you want me to do, like add doc comments to other files.

## My to-do list

- [x] Add comments to struct (let me know if this is good enough)
- [ ] Create mdx file that shows target markdown
- [ ] Define data required for intermediary file (JSON/YAML/whatever) -- I've added some ideas below

## Comment format

Here are few conventions followed in the talos example. I've used them in my comments, but we should discuss whether they are right for us:

- Tab before //.
- 3 spaces after //.
- Each sentence on its own line.
- Description is required

The talos example didn't have empty lines, but I've added them assuming it's Ok.

## Open questions

- When, if ever, do strings need to be quoted in the doc comments?
- Do we want to include a `name` key for future use (to support languages other than Go where name can't be derived programmatically)? What about a `type` key (same reason)?
- What types are valid? Integer, Float, Boolean, String, Null, Timestamp, List?
- In the talos example, there are no descriptions for lists of `values`. We have some settings where we will (I think) want to describe the values in detail. For example, in a topic like verification mode: https://www.elastic.co/guide/en/beats/filebeat/current/configuration-ssl.html, there are detailed descriptions.  I think we can still provide extended topics that describe how to use a bunch of settings together, but it would be better to have all the reference info in one place.
- What's the best way to handle doc links assuming they need to be versioned?

## Data we need in the intermediary file (YAML, JSON, whatever)

I think we need these keys/values starting out:

Key | Description | Required?
-- | -- | --
id* | Anchor ID | Y
name* | Setting name | Y
description | Description of the setting. May include multiple lines and markdown sections. | Y
shortdesc* | First sentence of the long description to be used in UI text strings and overview tables | Y
values | List of valid values | N
type* | Data type of the setting | Y
default | Default value of the setting | N
example | Name and reference to example | N

*We should be able to derive these programmatically (for Go programs)

We might need to add these keys/values later on to support other teams:

Key | Description
-- | -- 
cloud | Indicates whether the setting is available on ESS (true or false). Note that the cloud team already has a JSON file that identifies which user settings are supported on Cloud.
platform | Specifies a list of platforms on which the setting is supported.
setting-type | The type of Elastic setting (static or dynamic)
secure | Specifies that the setting is stored in the keystore (true or false)
stack-version | Specifies stack version when the setting was added
deprecated | Specifies if the field is deprecated (true or false)

dedemorton commented on pkg/typ/tls_config.go:
```diff
@@ -7,52 +7,95 @@ import (
 
 // TLSConfig is the interface used to configure a tcp client or server from a `Config`
+//
+//  examples:
```
this isn't hooked up to anything right now

ph replied:
I would say having example and test them are out of scope for the first version.

ph commented:
> - When, if ever, do strings need to be quoted in the doc comments?
I think you can just use quote directly, If I remember correctly the | allows you do to, example:
```yml
- yaml: |
    hello "work"
```


> - Do we want to include a `name` key for future use (to support languages other than Go where name can't be derived programmatically)? What about a `type` key (same reason)?

We can, I think this make it safe/explicit, also see in the struct the actual human option is in the tag of the field `yaml:...`

> - What types are valid? Integer, Float, Boolean, String, Null, Timestamp, List?
Integer, Float, Boolean, String,  we also have a Duration (1h, 5m) and all of them can be in a list.

> - In the talos example, there are no descriptions for lists of `values`. We have some settings where we will (I think) want to describe the values in detail. For example, in a topic like verification mode: https://www.elastic.co/guide/en/beats/filebeat/current/configuration-ssl.html, there are detailed descriptions.  I think we can still provide extended topics that describe how to use a bunch of settings together, but it would be better to have all the reference info in one place.

Yes, I think we need a way to link or included another types, ie Elasticsearch SSL settings are the same as Logstash SSL settings. I think this is what you mean? I am not sure yet if we can do it automatically or even if we should.

> - What's the best way to handle doc links assuming they need to be versioned?

Do we have use case? I would assume the doc generate from the same git commit are a vision in time. Do you want to refer in 8.x a doc about 7.x?

ph commented on pkg/typ/tls_config.go:
```diff
@@ -7,52 +7,95 @@ import (
+	//   description: |
+	//     Configures the SSL/TLS verification mode used during handshake.
+	//   values:
+	//     - full
```
Not sure how we could present the description of each of team, because I agree with you what `strict` mean without having a description about it? Maybe the answer here is to create the markdown and see how we can structure the data to correctly render the markdown?

ph replied:
@dedemorton Do you remember, that we had something like this for logstash, I think we can do better here because we don't have to run the code to parse it.

dedemorton replied:
Yes, I do remember! The Logstash plugin docs were complicated because the doc PRs had to be merged into the plugin repos, and the gem had to be published _before_ the docs could be built. Plus we were embedding actual asciidoc in the source code, and there was no validation to check for markup errors. We were breaking the doc build all the time, and the fixes took awhile because we had to bump the version and publish the gem to fix anything.

For this project, we've fewer complications. Plus we've come a long way wrt to testing PRs. If the doc comments don't build, then the doc build on the PR should fail. 

I also think we can keep the logic for generating the markdown files less complicated by writing all the content to a single file. The updated design of Next Docs (with a sticky, right-side mini TOC) will make it a lot easier for users to navigate through long files.

ph replied:
Yes, and we have an intermediate format, which we can easily validate.

ph commented:
@dedemorton one question I have, and I do not remember if we discussed it. Do the reference yml are generate from the same documentation?

ph commented:
Actually looking at my last comment maybe the reference is a bunch of link we follow. Looking at the sidero example it doesn't look super complicated. But maybe we were not as 'strict` or well-behaved when we wrote the configuration so we may not be able to use it for everything.

dedemorton commented:
@ph I think the Sidero Labs implementation allows users to generate commented reference YAML files, but I don't know if that's feasible for our project (you'd have to decide). It does seem pretty powerful, even if we generate the reference files ourselves and make them available (like we do now).

TBH, I'd be chuffed if we could generate Markdown files *and* something that can be used by the UI for text strings. Generating the reference files would be kind of an amazing bonus (but does seem like a lot more work).

It was my thought that having an intermediary file (like YAML or JSON) would make it easier for projects that want to maintain docs alongside the code, but don't want to embed it. And it would decouple the code comments somewhat from the target source format. Does that make sense? I wouldn't say it's a hard requirement, though, and maybe it's no worth the extra work. Not sure.

TBH honest, I don't completely understand the Sidero Labs code because I'm not a Go programmer, but it looks like they create an intermediary Go file.

dedemorton commented:
@ph OK, so I've added a bunch of comments and created a Markdown file that works in the new doc system. I think we need some more examples and other types of data we might want to capture, but maybe this is OK to begin with.

I'm pretty sure I've done weird things to the code, so you might need to fix it. :-) I'll be more careful when I work on real code, but I don't know how yaml unmarshaling (or wtf it's called) works.

I was confused how to deal with some things, so I added comments to the Markdown file.

The Next Docs syntax is described here, but it's incomplete: https://docs.elastic.dev/docs/syntax# 

The docs about building Next Docs are here: https://docs.elastic.dev/docs/setup

Here's the layout of the file right now (pretty basic):

![image](https://user-images.githubusercontent.com/14206422/159104559-73514eb8-5e18-463c-90fb-5cd6f7beb90d.png)

ph commented:
This is really great I was looking at the commits as you were making them. I will take a look more near the  end of next week. I will try to make something messy as fast as I can and clean it up.

Sent from my iPhone

> On Mar 18, 2022, at 11:15 PM, DeDe Morton ***@***.***> wrote:
> 
> ﻿
> @ph OK, so I've added a bunch of comments and created a Markdown file that works in the new doc system. I think we need some more examples and other types of data we might want to capture, but maybe this is OK to begin with.
> 
> I'm pretty sure I've done weird things to the code, so you might need to fix it. :-) I'll be more careful when I work on real code, but I don't know how yaml unmarshaling (or wtf it's called) works.
> 
> I was confused how to deal with some things, so I added comments to the Markdown file.
> 
> The Next Docs syntax is described here, but it's incomplete: https://docs.elastic.dev/docs/syntax#
> 
> The docs about building Next Docs are here: https://docs.elastic.dev/docs/setup
> 
> Here's the layout of the file right now (pretty basic):
> 
> 
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you were mentioned.

ph reviewed (approved):
I am going to merge and we can iterate.
